tinybytes/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+pub trait UnderlyingBytes: AsRef<[u8]> + Send + Sync + 'static {}
+
+/// Since the Bytes type is immutable, and UnderlyingBytes is `Send + Sync``, it is safe to share
+/// `Bytes` across threads.
+unsafe impl Send for Bytes {}
+unsafe impl Sync for Bytes {}
+
+impl Bytes {
+    /// Creates empty `Bytes`.
+    #[inline]
+    pub fn empty() -> Self {
+        Self::from_static(b"")
+    }
+
+    /// Creates `Bytes` from a static slice.
+    #[inline]
+    pub fn from_static(value: &'static [u8]) -> Self {
+        let slice: &[u8] = value;
+        Self {
+            ptr: slice.as_ptr(),
+            len: slice.len(),
+            bytes: None,
+        }
+    }
+
+    /// Creates `Bytes` from a slice, by copying.
+    pub fn copy_from_slice(data: &[u8]) -> Self {
+        Self::from_underlying(data.to_vec())
+    }
+
+    /// Returns the length of the `Bytes`.
+    #[inline]
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if the `Bytes` is empty.
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Returns a slice of self for the provided range.
+    ///
+    /// This will return a new `Bytes` handle set to the slice, and will not copy the underlying
+    /// data.
+    ///
+    /// This operation is `O(1)`.
+    ///
+    /// # Panics
+    ///
+    /// Slicing will panic if the range does not conform to  `start <= end` and `end <= self.len()`.
+    pub fn slice(&self, range: impl RangeBounds<usize>) -> Self {
+        use std::ops::Bound;
+
+        let len = self.len();
+
+        let start = match range.start_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n.checked_add(1).expect("range start overflow"),
+            Bound::Unbounded => 0,
+        };
+
+        let end = match range.end_bound() {
+            Bound::Included(&n) => n.checked_add(1).expect("range end overflow"),
+            Bound::Excluded(&n) => n,
+            Bound::Unbounded => len,
+        };
+
+        assert!(
+            start <= end,
+            "range start must not be greater than end: {:?} > {:?}",
+            start,
+            end,
+        );
+        assert!(
+            end <= len,
+            "range end must not be greater than length: {:?} > {:?}",
+            end,
+            len,
+        );
+
+        if end == start {
+            Bytes::empty()
+        } else {
+            self.safe_slice_ref(start, end)
+        }
+    }
+
+    /// Returns a slice of self that is equivalent to the given `subset`, if it is a subset.
+    ///
+    /// When processing a `Bytes` buffer with other tools, one often gets a
+    /// `&[u8]` which is in fact a slice of the `Bytes`, i.e. a subset of it.
+    /// This function turns that `&[u8]` into another `Bytes`, as if one had
+    /// called `self.slice()` with the range that corresponds to `subset`.
+    ///
+    /// This operation is `O(1)`.
```

Comment:
```suggestion
    /// This operation is `O(1)`.
    ///
    /// # Examples
    ///
    /// ```
    /// use tinybytes::Bytes;
    ///
    /// let bytes = Bytes::copy_from_slice(b"hello world");
    /// let subset = &bytes.as_ref()[0..5];
    /// let slice = bytes.slice_ref(subset).unwrap();
    /// assert_eq!(slice.as_ref(), b"hello");
    ///
    /// let subset = &bytes.as_ref()[6..11];
    /// let slice = bytes.slice_ref(subset).unwrap();
    /// assert_eq!(slice.as_ref(), b"world");
    ///
    /// let invalid_subset = b"invalid";
    /// assert!(bytes.slice_ref(invalid_subset).is_none());
    /// ```    
```